suite/tests/linux/threadset.c

```diff
@@ -0,0 +1,353 @@
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of VMware, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Tests resuming from check_wait_at_safe_spot => thread_set_self_context,
+ * triggered by another thread flushing (causing a synchall). Test based on
+ * linux.sigcontext.
+ */
+
+#include "tools.h"
+#include "thread.h"
+
+/* we want the latest defs so we can get at ymm state */
+#include "../../../core/unix/include/sigcontext.h"
+#include <stdio.h>
+#include <unistd.h>
+#include <signal.h>
+#include <ucontext.h>
+#include <errno.h>
+#include <stddef.h>
+
+/* For sharing NUM_*_REGS constants. */
+#include "../api/detach_state_shared.h"
+
+#define INTS_PER_XMM 4
+#define INTS_PER_YMM 8
+#define INTS_PER_ZMM 16
+
+NOINLINE void
+dummy2()
+{
+    for (int i = 0; i < 10; i++) {
+        asm volatile("add %rdi, %rdi");
+    }
+}
+
+void *
+thread()
+{
+    for (int i = 0; i < 100000; i++) {
+        dummy2();
+    }
+}
+
+int
+main(int argc, char *argv[])
+{
+    int buf[INTS_PER_XMM * NUM_SIMD_SSE_AVX_REGS];
+    char *ptr = (char *)buf;
+    int i, j;
+
+    /* this test deliberately uses write() instead of the other libc calls, since those
+     * appeared to cause crashes (likely due to us accidentically triggering the xmm
+     * saving bug :/).
+     */
+
+    const char start_msg[] = "Starting test.\n";
+    const char saving_regs[] = "Saving regs.\n";
+    const char before_msg[] = "Before synchall loop.\n";
+    const char after_msg[] = "After synchall loop.\n";
+    const char assertion_failed[] = "Assertion failed.\n";
+    const char all_done[] = "All done.\n";
+
+    write(2, start_msg, sizeof(start_msg) - 1);
+
+    thread_t flusher = create_thread(thread, NULL);
+    sleep(1);
```

Comment:
Do we need this?  If so, can we instead use conditional vars (see our condvar.h) to synchronize with the thread to ensure it's created?  Generally, relying on timing is going to make this flaky on slow VM's in automated testing.